assets/js/components/KeyMetrics/ConnectGA4CTAWidget.test.js

```diff
@@ -0,0 +1,215 @@
+/**
+ * ConnectGA4CTAWidget component tests.
+ *
+ * Site Kit by Google, Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Internal dependencies
+ */
+import ConnectGA4CTAWidget from './ConnectGA4CTAWidget';
+import {
+	CORE_USER,
+	KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
+	KM_ANALYTICS_LOYAL_VISITORS,
+	KM_ANALYTICS_NEW_VISITORS,
+	KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
+	KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
+	KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT,
+} from '../../googlesitekit/datastore/user/constants';
+import {
+	render,
+	createTestRegistry,
+	provideModules,
+	provideKeyMetrics,
+	provideUserAuthentication,
+	provideUserCapabilities,
+} from '../../../../tests/js/test-utils';
+import { CORE_WIDGETS } from '../../googlesitekit/widgets/datastore/constants';
+import { AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY } from '../../googlesitekit/widgets/default-areas';
+import { CONTEXT_MAIN_DASHBOARD_KEY_METRICS } from '../../googlesitekit/widgets/default-contexts';
+
+describe( 'ConnectGA4CTAWidget', () => {
+	let registry;
+
+	const DISMISSED_ITEM_KEY = 'key-metrics-connect-ga4-cta-widget';
+
+	const Widget = ( { children } ) => <div>{ children }</div>;
+	const WidgetNull = () => <div>NULL</div>;
+
+	beforeEach( () => {
+		registry = createTestRegistry();
+
+		provideUserAuthentication( registry );
+		provideUserCapabilities( registry );
+		provideKeyMetrics( registry );
+
+		registry.dispatch( CORE_USER ).receiveGetDismissedItems( [] );
+		registry.dispatch( CORE_USER ).receiveIsUserInputCompleted( true );
+
+		provideModules( registry, [
+			{
+				slug: 'analytics-4',
+				active: false,
+				connected: false,
+			},
+		] );
+	} );
+
+	it( 'should not render when the widget is dismissed', async () => {
+		registry
+			.dispatch( CORE_USER )
+			.receiveGetDismissedItems( [ DISMISSED_ITEM_KEY ] );
+
+		const { container, waitForRegistry } = render(
+			<ConnectGA4CTAWidget Widget={ Widget } WidgetNull={ WidgetNull } />,
+			{
+				registry,
+			}
+		);
+		await waitForRegistry();
+		expect( container ).toHaveTextContent( 'NULL' );
+	} );
+
+	it( 'should not render if user input is not completed', async () => {
+		global._googlesitekitUserData.isUserInputCompleted = false;
+		await registry
+			.dispatch( CORE_USER )
+			.receiveIsUserInputCompleted( false );
```

Comment:
It seems unnecessary that we'd need to both set the global and receive the same value directly. We should be able to use the datastore exclusively – using the global in a test should only be done in the tests that are intended to exercise the resolver which uses it.

Reply:
This test case has been removed. However, as far as I can remember, I was encountering an error when not setting the global, which is odd and needed to be investigated further.

Reply:
This test case has been removed. However, as far as I can remember, I was encountering an error when not setting the global, which is odd and needed to be investigated further.

Reply:
Thanks! The problem seems to be here in the resolver where it only returns early if user input has been completed https://github.com/google/site-kit-wp/blob/9507eaf5369ebbdac51858154b49e3fc43a9af92/assets/js/googlesitekit/datastore/user/user-info.js#L267-L269

This should be changed to return early any time the value is not `undefined`.

So even though we're providing the value directly, the resolver is still trying to get the value (and receive it again) from the global.

Reply:
Excellent catch. Understood and updated, thank you!